### PR TITLE
Bools are IO-shareable, but only for builtins

### LINKED ISSUE #1244: update IO-shareable to include bool
#1242 made builtin(front_facing) a boolean.
So we have to make bool IO-shareable.

Note that Vulkan allows bool for builtin input/output variables, but not for user inputs/outputs.

From [15.1.2 User-defined Variable Interface](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#interfaces-iointerfaces-user)

> The non-built-in variables listed by OpEntryPoint with the Input or Output storage class form the user-defined variable interface. These must have SPIR-V numerical types or, recursively, composite types of such types.

### PR DESCRIPTION
**Pull request by dneto0**
Fixes #1244

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -499,6 +499,10 @@ The following kinds of values must be of IO-shareable type:
 * Values accepted as inputs from an upstream pipeline stage.
 * Values written as output for downstream processing in the pipeline, or to an output attachment.
 
+Note: Only built-in pipeline inputs may have a boolean type.
+A user data attribute input or output must not be of [=bool=] type or contain a [=bool=] type.
```
Comment (dj2) [inline wgsl/index.bs:503]: This reads strangely. Maybe `A user input or output data attribure ..`?

File: wgsl/index.bs [hunk 2]
```diff
@@ -534,8 +538,9 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
          of this member.
 </table>
 
-Note: An [=IO-shareable=] type would also be host-shareable if it and its subtypes have
-the approporate [=stride=] and [=offset=] attributes.
+Note: An [=IO-shareable=] type *T* would also be host-shareable if *T* and its subtypes have
+appropriate [=stride=] and [=offset=] attributes, and if *T* is not [=bool=] and did
```
Comment (dj2) [inline wgsl/index.bs:542]: nit: did -> does ?

**Review by kvark — COMMENTED** [review]
Why can't the user I/O data have bools? Do we have the relevant investigation somewhere?

**Comment by kainino0x** [comment]
OpTypeBool

Declare the Boolean type. Values of this type can only be either true or false. There is no physical size or bit pattern defined for these values. If they are stored (in conjunction with OpVariable), they must only be used with logical addressing operations, not physical, and only with non-externally visible shader Storage Classes: Workgroup, CrossWorkgroup, Private, Function, Input, and Output.

**Comment by kainino0x** [comment]
And you probably wouldn't want to have WGSL give them a bit pattern (i.e. define bools as translating to OpTypeInt) unless you had 1-byte storage (not possible in core), otherwise they would waste bandwidth.

**Comment by kvark** [comment]
@kainino0x what you posted actually agrees with my question and contradicts @dneto0 's PR: it says Input and Output are valid storage classes for bools. Where does it say that Input and Output bools are only allowed for builtins?

**Comment by grorg** [comment]
Discussed at the [2020-12-01 meeting](https://docs.google.com/document/d/1BGljZlq1ZY1DTZS_MHxTxfu6rF0j87uUrPL7RHArpLA/). 

Resolution: Accept with the update. @kvark will file a post-MVP issue on relaxing the rule.

**Comment by dneto0** [comment]
I think I've addressed your feedback.  PTAL @dj2

**Comment by dneto0** [comment]
@kvark  asked

>  Where does it say that Input and Output bools are only allowed for builtins?

OpTypeBool can't be on the pipeline input or output boundary because OpTypeBool does not have a concrete bit representation.

There's this note in the SPIR-V spec:

>  Khronos Issue #363: OpTypeBool can be used in the Input and Output storage classes, but the client APIs still only allow built-in Boolean variables (e.g. FrontFacing), not user variables. 

(I'm looking for a more definitive statement in Vulkan...)

**Comment by dneto0** [comment]
Ah, here it is. From [15.1.2 User-defined Variable Interface](https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#interfaces-iointerfaces-user)

> The non-built-in variables listed by OpEntryPoint with the Input or Output storage class form the user-defined variable interface. These must have SPIR-V numerical types or, recursively, composite types of such types.

That excludes bool.

**Comment by dneto0** [comment]
Discussed in [2021-01-19 meeting](https://docs.google.com/document/d/1xgCv1_xzo6kRmCfwOtnPd2j_T4unot71BOnlF-gOlIE/edit#heading=h.oubgp0qo6qo5)

- @kvark suggested allowing bool on input and output, and polyfilling for Vulkan, with 0 interpreted as false, non-zero as true.

**Comment by dneto0** [comment]
Ok, so there is agreement on the content. Now this PR just needs editorial review.

**Review by dj2 — APPROVED** [review]

**Pull request merged**